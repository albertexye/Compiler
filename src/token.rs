use serde::Serialize;
use std::collections::HashMap;
use std::ops::Sub;
#[cfg(test)]
use std::sync::{LazyLock, Mutex};

#[derive(Debug, PartialEq, Eq, Clone, Copy)]
pub(crate) enum TokenType {
    Comma,
    Semicolon,
    Colon,
    DoubleColon,
    Dot,
    OpenParen,
    CloseParen,
    OpenBrace,
    CloseBrace,
    OpenBracket,
    CloseBracket,
    Plus,
    PlusEq,
    Minus,
    MinusEq,
    Mul,
    MulEq,
    Div,
    DivEq,
    Modulo,
    ModuloEq,
    LeftShift,
    LeftShiftEq,
    RightShift,
    RightShiftEq,
    BitAnd,
    BitAndEq,
    BitOr,
    BitOrEq,
    BitXor,
    BitXorEq,
    BitNot,
    LogicalAnd,
    LogicalOr,
    LogicalNot,
    Eq,
    NotEq,
    Gt,
    Ge,
    Lt,
    Le,
    Assign,
    ReturnType,
    MatchCase,
    If,
    Else,
    Match,
    While,
    For,
    Break,
    Continue,
    Return,
    Fn,
    Let,
    Var,
    Struct,
    Enum,
    Union,
    Pub,
    Prv,
    Mod,
    Module,
    Import,
    Use,
}

#[derive(Debug, PartialEq, Clone)]
pub(crate) enum Literal {
    UInt(u64),
    Int(i64),
    Float(f64),
    String(String),
}

#[derive(Debug, PartialEq, Clone)]
pub(crate) enum TokenValue {
    Identifier(SymbolId),
    Literal(Literal),
    Keyword(TokenType),
}

#[derive(Debug, PartialEq, Clone, Copy, Serialize)]
pub(crate) struct TokenSpan {
    pub(crate) line: usize,
    pub(crate) column: usize,
    pub(crate) index: usize,
    pub(crate) size: usize,
}

#[derive(Debug, PartialEq, Clone)]
pub(crate) struct Token {
    pub(crate) value: TokenValue,
    pub(crate) span: TokenSpan,
}

impl Sub for TokenSpan {
    type Output = TokenSpan;

    fn sub(self, other: TokenSpan) -> TokenSpan {
        std::debug_assert!(self.index + self.size >= other.index);
        TokenSpan {
            line: self.line,
            column: self.column,
            index: self.index,
            size: self.index + self.size - other.index,
        }
    }
}

const TOKEN_TYPES_STR: [&str; 64] = [
    ",", ";", ":", "::", ".", "(", ")", "[", "]", "{", "}", "+", "+=", "-", "-=", "*", "*=", "/",
    "/=", "%", "%=", "<<", "<<=", ">>", ">>=", "&", "&=", "|", "|=", "^", "^=", "~", "and", "or",
    "!", "==", "!=", ">", ">=", "<", "<=", "=", "->", "=>", "if", "else", "match", "while", "for",
    "break", "continue", "return", "fn", "let", "var", "struct", "enum", "union", "pub", "prv",
    "mod", "module", "import", "use",
];

const TOKEN_TYPES_ENUM: [TokenType; 64] = [
    TokenType::Comma,
    TokenType::Semicolon,
    TokenType::Colon,
    TokenType::DoubleColon,
    TokenType::Dot,
    TokenType::OpenParen,
    TokenType::CloseParen,
    TokenType::OpenBrace,
    TokenType::CloseBrace,
    TokenType::OpenBracket,
    TokenType::CloseBracket,
    TokenType::Plus,
    TokenType::PlusEq,
    TokenType::Minus,
    TokenType::MinusEq,
    TokenType::Mul,
    TokenType::MulEq,
    TokenType::Div,
    TokenType::DivEq,
    TokenType::Modulo,
    TokenType::ModuloEq,
    TokenType::LeftShift,
    TokenType::LeftShiftEq,
    TokenType::RightShift,
    TokenType::RightShiftEq,
    TokenType::BitAnd,
    TokenType::BitAndEq,
    TokenType::BitOr,
    TokenType::BitOrEq,
    TokenType::BitXor,
    TokenType::BitXorEq,
    TokenType::BitNot,
    TokenType::LogicalAnd,
    TokenType::LogicalOr,
    TokenType::LogicalNot,
    TokenType::Eq,
    TokenType::NotEq,
    TokenType::Gt,
    TokenType::Ge,
    TokenType::Lt,
    TokenType::Le,
    TokenType::Assign,
    TokenType::ReturnType,
    TokenType::MatchCase,
    TokenType::If,
    TokenType::Else,
    TokenType::Match,
    TokenType::While,
    TokenType::For,
    TokenType::Break,
    TokenType::Continue,
    TokenType::Return,
    TokenType::Fn,
    TokenType::Let,
    TokenType::Var,
    TokenType::Struct,
    TokenType::Enum,
    TokenType::Union,
    TokenType::Pub,
    TokenType::Prv,
    TokenType::Mod,
    TokenType::Module,
    TokenType::Import,
    TokenType::Use,
];

const _: () = assert!(TOKEN_TYPES_STR.len() == TOKEN_TYPES_ENUM.len());

#[cfg_attr(not(test), derive(Serialize))]
#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
pub(crate) struct SymbolId(usize);

pub(crate) struct SymbolTable {
    counter: SymbolId,
    table: HashMap<String, SymbolId>,
    reverse: Option<Vec<String>>,
}

impl SymbolTable {
    pub(crate) fn new() -> SymbolTable {
        let mut sym_table = SymbolTable {
            counter: SymbolId(0),
            table: HashMap::new(),
            reverse: None,
        };
        for keyword in TOKEN_TYPES_STR {
            sym_table
                .table
                .insert(keyword.to_string(), sym_table.counter);
            sym_table.counter.0 += 1;
        }
        sym_table
    }

    pub(crate) fn insert(&mut self, token: String) -> SymbolId {
        std::debug_assert!(self.reverse.is_none());
        if self.table.contains_key(&token) {
            self.table[&token]
        } else {
            let id = self.counter;
            self.table.insert(token, self.counter);
            self.counter.0 += 1;
            id
        }
    }

    pub(crate) fn search(&self, token: &str) -> Option<SymbolId> {
        std::debug_assert!(self.reverse.is_none());
        if self.table.contains_key(token) {
            Some(self.table[token])
        } else {
            None
        }
    }

    pub(crate) fn is_keyword(id: &SymbolId) -> bool {
        id.0 < TOKEN_TYPES_STR.len()
    }

    pub(crate) fn get_keyword(id: &SymbolId) -> Option<TokenType> {
        if !SymbolTable::is_keyword(id) {
            None
        } else {
            Some(TOKEN_TYPES_ENUM[id.0])
        }
    }

    pub(crate) fn reverse_lookup(&mut self, id: SymbolId) -> String {
        let rev = match self.reverse.as_ref() {
            Some(rev) => rev,
            None => {
                let mut reverse = vec![String::new(); self.counter.0];
                let table = std::mem::take(&mut self.table);
                for (sym, id) in table.into_iter() {
                    reverse[id.0] = sym;
                }
                self.reverse = Some(reverse);
                self.reverse.as_ref().unwrap()
            }
        };
        assert!(id.0 < rev.len());
        rev[id.0].clone()
    }
}

#[cfg(test)]
static SYMBOL_CONTEXT: LazyLock<Mutex<SymbolTable>> =
    LazyLock::new(|| Mutex::new(SymbolTable::new()));

#[cfg(test)]
pub(crate) fn set_symbol_context(symbol_table: SymbolTable) {
    let mut guard = SYMBOL_CONTEXT.lock().unwrap();
    *guard = symbol_table;
}

#[cfg(test)]
impl Serialize for SymbolId {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        let mut guard = SYMBOL_CONTEXT.lock().unwrap();
        serializer.serialize_str(&guard.reverse_lookup(*self))
    }
}
