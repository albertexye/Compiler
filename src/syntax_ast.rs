use std::{
    collections::{HashMap, HashSet},
    path::PathBuf,
};

use crate::token::TokenSpan;

pub(crate) struct Ast {
    pub(crate) entry: PathBuf,
    pub(crate) modules: HashMap<PathBuf, Module>,
}

pub(crate) struct Module {
    pub(crate) name: String,
    pub(crate) files: HashMap<String, File>,
}

pub(crate) struct File {
    pub(crate) name: String,
    pub(crate) module: String,
    pub(crate) imports: HashSet<String>,
    pub(crate) globals: HashMap<String, Scope<Declaration>>,
    pub(crate) functions: HashMap<String, Scope<Function>>,
    pub(crate) types: HashMap<String, Scope<TypeDef>>,
}

pub(crate) type Name = Vec<String>;

pub(crate) enum Visibility {
    Public,
    Private,
    Module,
}

pub(crate) struct Scope<T> {
    pub(crate) visibility: Visibility,
    pub(crate) value: T,
}

pub(crate) struct TypeDef {
    pub(crate) name: String,
    pub(crate) body: TypeDefBody,
    pub(crate) span: TokenSpan,
}

pub(crate) enum TypeDefBody {
    Struct(HashMap<String, TypeAnnot>),
    Enum(HashMap<String, u64>),
    Union(HashMap<String, TypeAnnot>),
    Alias(TypeAnnot),
}

pub(crate) enum Statement {
    Declaration(Declaration),
    Assignment(Assignment),
    Expression(Expression),
    Loop(Loop),
    Continue(TokenSpan),
    Break(TokenSpan),
    Conditional(Conditional),
    Match(Match),
    Return(Expression),
}

pub(crate) struct TypeAnnot {
    pub(crate) base: Name,
    pub(crate) modifiers: Vec<TypeModifier>,
    pub(crate) span: TokenSpan,
}

pub(crate) struct TypeModifier {
    pub(crate) mutable: bool,
    pub(crate) typ: TypeModifierType,
}

pub(crate) enum TypeModifierType {
    Pointer,
    Slice,
    Array(u64),
}

pub(crate) struct Function {
    pub(crate) name: String,
    pub(crate) arguments: Vec<FunctionArg>,
    pub(crate) return_type: Option<TypeAnnot>,
    pub(crate) body: Vec<Statement>,
    pub(crate) span: TokenSpan,
}

pub(crate) struct FunctionArg {
    pub(crate) name: String,
    pub(crate) typ: TypeAnnot,
    pub(crate) span: TokenSpan,
}

pub(crate) struct Declaration {
    pub(crate) name: String,
    pub(crate) mutable: bool,
    pub(crate) typ: TypeAnnot,
    pub(crate) value: Expression,
    pub(crate) span: TokenSpan,
}

pub(crate) struct Expression {
    pub(crate) value: ExpressionValue,
    pub(crate) span: TokenSpan,
}

pub(crate) enum ExpressionValue {
    Binary(Binary),
    Unary(Unary),
    Call(Call),
    Literal(Literal),
    Identifier(Name),
}

pub(crate) struct Binary {
    pub(crate) left: Box<Expression>,
    pub(crate) right: Box<Expression>,
    pub(crate) op: BinaryOp,
}

pub(crate) enum BinaryOp {
    Plus,
    Minus,
    Mul,
    Div,
    Mod,
    LeftShift,
    RightShift,
    BitAnd,
    BitOr,
    BitXor,
    Gt,
    Ge,
    Lt,
    Le,
    Eq,
    NotEq,
    LogicalAnd,
    LogicalOr,
    Indexing,
    FieldAccess,
}

pub(crate) struct Unary {
    pub(crate) operand: Box<Expression>,
    pub(crate) op: UnaryOp,
}

pub(crate) enum UnaryOp {
    LogicalNot,
    BitNot,
    Dereference,
    AddressOf,
    Negate,
}

pub(crate) struct Call {
    pub(crate) function: Box<Expression>,
    pub(crate) args: Vec<Expression>,
}

pub(crate) enum Literal {
    String(String),
    UInt(u64),
    Int(i64),
    Float(f64),
    Bool(bool),
    Array(Vec<Expression>),
    Struct(HashMap<String, Expression>),
}

pub(crate) struct Assignment {
    pub(crate) left: Expression,
    pub(crate) right: Expression,
    pub(crate) typ: AssignmentType,
    pub(crate) span: TokenSpan,
}

pub(crate) enum AssignmentType {
    Assign,
    Plus,
    Minus,
    Mul,
    Div,
    Mod,
    LeftShift,
    RightShift,
    BitAnd,
    BitOr,
    BitXor,
}

pub(crate) struct ConditionalBranch {
    pub(crate) condition: Expression,
    pub(crate) body: Vec<Statement>,
}

pub(crate) struct Conditional {
    pub(crate) if_branch: ConditionalBranch,
    pub(crate) elif_branches: Vec<ConditionalBranch>,
    pub(crate) else_branch: Option<Vec<Statement>>,
}

pub(crate) struct Match {
    pub(crate) value: Expression,
    pub(crate) cases: Vec<ConditionalBranch>,
    pub(crate) default: Option<Vec<Statement>>,
}

pub(crate) struct Loop {
    pub(crate) init: Option<Declaration>,
    pub(crate) condition: Option<Expression>,
    pub(crate) update: Vec<Statement>,
    pub(crate) body: Vec<Statement>,
}
