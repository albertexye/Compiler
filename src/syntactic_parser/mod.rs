use std::io;

use crate::syntax_ast;
use crate::syntax_ast::{Name, Statement};
use crate::token;
use crate::token::{Token, TokenType, TokenValue};

mod assignment;
mod conditional;
mod declaration;
mod expression;
mod file;
mod function;
mod r#loop;
mod r#match;
mod r#return;
mod statement;
mod type_annotation;
mod type_definition;

#[derive(Debug)]
pub(crate) enum ErrorType {
    Name,
    Module,
    Import,
    LineEnd,
    TypeDefinition,
    Declaration,
    TypeAnnotation,
    Expression,
    Statement,
    Conditional,
    Function,
    Match,
    ModuleSchema(serde_json::Error),
    ModuleNotFound,
    Io(io::Error),
}

#[derive(Debug)]
pub(crate) struct Error {
    typ: ErrorType,
    msg: String,
    token: Option<Token>,
}

pub struct SyntacticParser {
    tokens: Vec<Token>,
    index: usize,
}

impl SyntacticParser {
    pub(crate) fn parse(
        tokens: Vec<Token>,
        filename: &str,
        module_name: &str,
    ) -> Result<syntax_ast::File, Error> {
        let mut parser = SyntacticParser {
            tokens,
            index: 0usize,
        };
        parser.parse_file(filename, module_name)
    }

    fn peek(&self) -> Option<Token> {
        self.tokens.get(self.index).cloned()
    }

    fn back(&self) -> Token {
        self.tokens[self.index - 1].clone()
    }

    fn error(&self, error_type: ErrorType, message: &str) -> Error {
        Error {
            typ: error_type,
            msg: message.to_string(),
            token: self.peek(),
        }
    }

    fn expect_token(&self, error_type: ErrorType, message: &str) -> Result<Token, Error> {
        match self.peek() {
            Some(token) => Ok(token),
            None => Err(self.error(error_type, message)),
        }
    }

    fn expect_identifier(&self, error_type: ErrorType, message: &str) -> Result<String, Error> {
        let token = match self.peek() {
            Some(token) => token,
            None => return Err(self.error(error_type, message)),
        };
        match token.value {
            TokenValue::Identifier(id) => Ok(id),
            _ => Err(self.error(error_type, message)),
        }
    }

    fn expect_keyword(
        &self,
        kw: TokenType,
        error_type: ErrorType,
        message: &str,
    ) -> Result<(), Error> {
        if !self.is_keyword(kw) {
            Err(self.error(error_type, message))
        } else {
            Ok(())
        }
    }

    fn advance(&mut self) {
        self.index += 1;
    }

    fn is_keyword(&self, keyword: TokenType) -> bool {
        let Some(token) = self.peek() else {
            return false;
        };
        match token.value {
            TokenValue::Keyword(kwd) => kwd == keyword,
            _ => false,
        }
    }

    fn end_line(&mut self) -> Result<(), Error> {
        if !self.is_keyword(TokenType::Semicolon) {
            Err(self.error(ErrorType::LineEnd, "`;` expected at end of line"))
        } else {
            self.advance();
            Ok(())
        }
    }

    fn is_identifier(&self) -> Option<String> {
        let token = self.peek()?;
        match token.value {
            TokenValue::Identifier(id) => Some(id),
            _ => None,
        }
    }

    fn is_mutable(&self) -> Result<bool, Error> {
        if self.is_keyword(TokenType::Let) {
            Ok(false)
        } else if self.is_keyword(TokenType::Var) {
            Ok(true)
        } else {
            Err(self.error(
                ErrorType::TypeAnnotation,
                "Type annotations must specify mutability",
            ))
        }
    }

    fn is_uint(&self) -> Option<u64> {
        let token = self.peek()?;
        match token.value {
            TokenValue::Literal(token::Literal::UInt(uint)) => Some(uint),
            _ => None,
        }
    }

    fn parse_block(&mut self) -> Result<Vec<Statement>, Error> {
        if !self.is_keyword(TokenType::OpenBracket) {
            return Err(self.error(ErrorType::Conditional, "Expected contional body"));
        }
        self.advance();
        let mut statements = Vec::new();
        while !self.is_keyword(TokenType::CloseBracket) {
            statements.push(self.parse_statement()?);
        }
        self.advance();
        Ok(statements)
    }

    fn parse_name(&mut self) -> Result<Name, Error> {
        std::debug_assert!(self.is_identifier().is_some());
        let mut name = Vec::new();
        name.push(self.is_identifier().unwrap());
        self.advance();
        while self.is_keyword(TokenType::DoubleColon) {
            self.advance();
            if let Some(id) = self.is_identifier() {
                name.push(id);
                self.advance();
            } else {
                return Err(self.error(ErrorType::Expression, "Expected identifier"));
            }
        }
        Ok(name)
    }
}
