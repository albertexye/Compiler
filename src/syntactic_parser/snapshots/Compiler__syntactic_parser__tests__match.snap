---
source: src/syntactic_parser/mod.rs
expression: ast
---
name: test
module: test_match
imports:
  - std
globals: {}
functions:
  is_true:
    visibility: Public
    value:
      name: is_true
      arguments:
        - name: cond
          typ:
            base:
              - bool
            modifiers: []
            span:
              line: 5
              column: 22
              index: 54
              size: 4
          span:
            line: 5
            column: 22
            index: 54
            size: 10
      return_type:
        base:
          - bool
        modifiers: []
        span:
          line: 5
          column: 31
          index: 63
          size: 4
      body:
        - Match:
            value:
              value:
                Identifier:
                  - cond
              span:
                line: 6
                column: 12
                index: 81
                size: 4
            cases:
              - condition:
                  value:
                    Identifier:
                      - "true"
                  span:
                    line: 7
                    column: 9
                    index: 97
                    size: 4
                body:
                  - Return:
                      value:
                        Identifier:
                          - "true"
                      span:
                        line: 7
                        column: 26
                        index: 114
                        size: 4
              - condition:
                  value:
                    Identifier:
                      - "false"
                  span:
                    line: 8
                    column: 9
                    index: 130
                    size: 5
                body:
                  - Return:
                      value:
                        Identifier:
                          - "false"
                      span:
                        line: 8
                        column: 27
                        index: 148
                        size: 5
            default:
              - Expression:
                  value:
                    Call:
                      function:
                        value:
                          Identifier:
                            - std
                            - print
                        span:
                          line: 9
                          column: 21
                          index: 177
                          size: 10
                      args:
                        - value:
                            Literal:
                              String: Never happends
                          span:
                            line: 9
                            column: 27
                            index: 183
                            size: 16
                  span:
                    line: 9
                    column: 43
                    index: 199
                    size: 18
      span:
        line: 5
        column: 8
        index: 40
        size: 7
types: {}
